### ISSUE TITLE
vkEnumerateInstanceExtensionProperties cannot be called from CreateInstance in a layer

### ISSUE THREAD

**Issue by Themaister** [KhronosGroup/Vulkan-LoaderAndValidationLayers#2064]
I am trying to write a Vulkan layer which injects functionality which requires particular extensions from an ICD.

The application might not enable those extensions, so I am trying to add those extensions in the layer's CreateInstance when going down the chain.

To do this robustly, I need to query that the extensions actually exist before calling pfnNextGetInstanceProcAddr. However, querying vkEnumerateInstanceExtensionProperties fails, yet vkCreateInstance works. This seems a bit odd.

```
	auto fpGetInstanceProcAddr = chainInfo->u.pLayerInfo->pfnNextGetInstanceProcAddr;

	auto fpEnumerateExts = reinterpret_cast<PFN_vkEnumerateInstanceExtensionProperties>(
			fpGetInstanceProcAddr(nullptr, "vkEnumerateInstanceExtensionProperties"));
	if (!fpEnumerateExts)
		return VK_ERROR_INITIALIZATION_FAILED;
```

**Comment by MarkY-LunarG** [comment]
This is currently "as designed".  The original intent was this information was queried by the desktop loader from the layer JSON files.  However, we are investigating the usefulness of this.

**Comment by Themaister** [comment]
Similar for vkEnumerateDeviceExtensionProperties in CreateDevice.

**Comment by Themaister** [comment]
The current workaround is dlsym(RTLD_DEFAULT, ...)

**Comment by Themaister** [comment]
Actually, it's not ... vkEnumerateInstanceExtensions deadlocks in libvulkan.so. Probably some global lock is taken in vkCreateInstance and in vkEnumerateInstanceExtensions as well.

**Comment by MarkY-LunarG** [comment]
I'm not sure why dlsym changes would make it work.  The `vkEnumerateInstanceExtensionProperties` extension should be implemented by the loader and ICDs only and you shouldn't attempt to try to call the ICD function directly.  We're looking into a more proper way to do this.  However, I don't see why `vkEnumerateDeviceExtensionProperties` shouldn't work by default.  I've assigned Lenny this issue, and he can look into more detail.

**Comment by Themaister** [comment]
Yeah, dlsym(RTLD_DEFAULT) actually didn't work, so please disregard that.

What I tried was dlopen("libvulkan.so") directly, then call into vkEnumerateInstanceExtensionProperties, which deadlocked.

**Comment by lenny-lunarg** [comment]
Sorry for the delay on this issue. I still haven't looked into exactly what this failure is, but I am working on a patch that will allow implicit layers to intercept functions like `vkEnumerateInstanceExtensionProperties` which are called before an instance is created. This is done to allow implicit layers to hide things that they won't support. This won't directly address your problem, but it is related. I'll look into this specific issue in more detail shortly.

So the main thing here is I'm not ignoring this issue, I just haven't had a chance to look into this, but I will shortly.
